core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/AttributeComponent.scala

```diff
@@ -310,22 +319,47 @@ trait AttributeComponent {
       filter(_.id inSetBind attributeRecords.map(_.id)).delete
     }
 
-    object AlterAttributesUsingScratchTableQueries extends RawSqlQuery {
-      val driver: JdbcDriver = AttributeComponent.this.driver
-      /*!CAUTION! - do not drop the ownerIds clause at the end or it will delete every attribute in the table!
+    // for DB performance reasons, it's necessary to split "save attributes" into 3 steps:
+    // - delete attributes which are not present in the new set
+    // - insert attributes which are not present in the existing set
+    // - update attributes which are common to both sets
+    // Delete and Insert use standard Slick actions, but Update uses a scratch table
+    // (AlterAttributesUsingScratchTableQueries) for DB performance reasons.
+
+    // we use set operations (filter, contains, partition) to split the attributes,
+    // which uses the element type's equals method.  Ideally we would use AttributeRecord.equals,
+    // but we only want to use a subset of the fields: the primary key.
+    // AttributeRecordPrimaryKey encapsulates only those fields.
+
+    case class AttributeRecordPrimaryKey(ownerId: OWNER_ID, namespace: String, name: String, listIndex: Option[Int]) {
+      override def equals(o: Any) = o match {
```

Comment:
I take that "does not really matter" comment back. You are using these as keys in a Map and there likely to be a lot of them so a good hashCode function probably matters quite a bit. Back in my java days I would say you should also be intelligently sizing your map... don't know if that matters in scala or not. Perhaps profiling this for a large upload is a good idea.